synapse/handlers/federation.py

```diff
@@ -1667,16 +1667,52 @@ async def on_send_join_request(self, origin: str, pdu: EventBase) -> JsonDict:
         # would introduce the danger of backwards-compatibility problems.
         event.internal_metadata.send_on_behalf_of = origin
 
-        context = await self._handle_new_event(origin, event)
+        # Calculate the event context.
+        context = await self._prep_event(
+            origin, event, state=None, auth_events=None, backfilled=False
+        )
+
+        # Get the current state at the to-be created event.
+        prev_state_ids = await context.get_prev_state_ids()
+
+        # Check if the user is already in the room or invited to the room.
+        user_id = event.state_key
+        prev_member_event_id = prev_state_ids.get((EventTypes.Member, user_id), None)
+        newly_joined = True
+        is_invite = False
+        if prev_member_event_id:
+            prev_member_event = await self.store.get_event(prev_member_event_id)
+            newly_joined = prev_member_event.membership != Membership.JOIN
+            is_invite = prev_member_event.membership == Membership.INVITE
```

Comment:
Much of this logic (and the if-statement below) is now duplicated between this and local joins. I wonder if we should instead update `can_join_without_invite` to raise a `SynapseError` and try to push some of this logic in there? The local join logic has a couple of interleaved bits though I think.

Reply:
I don't think it would be the end of the world to have this logic in both `_local_membership_update` and `can_join_without_invite`, to save us from duplicating the `if newly_joined and not is_invite` condition. I don't think it matters too much though.